### ISSUE TITLE
Method declarations object literals don't remove self parameter

### ISSUE THREAD

**Issue by GlassBricks** [TypeScriptToLua/TypeScriptToLua#1081]
When creating an object literal implementing an interface, methods (shorthand functions) don't have `self` removed if the corresponding interface function has noSelf.
Same with mapped type literals.

Example here:
https://typescripttolua.github.io/play#code/FDCWDsBcFMCcDMCGBjaACAYgV3MyoB7cAYSMkQjjQG9g1614CCAuNACkgAtQBnFgG4FQAEwA0aRG16RYEAOYBKALwA+IaLoMARolhtOPfhvGTpshSvXCRW+roBeB7n0E2JUtDLnglak8AAviDIRDJoALYAnti4pFAU4HBssXiEJGSJVMo0dozM7IiKtAylaKHgvAQANtAAdNUE8oWKeYFiebr6jDhpRC0lZfQVVbUNTS1tHaWOBkVqg0MjNfWNzUVtQSDRqfHklLB1TATsAEQ8p607vXtZh11nF1cxN5kHdY6PoJdAA

**Comment by GlassBricks** [comment]
Another possibly related issue... when I first load the playground link above, an error is reported. But as soon as I edit the code the error disapears.

The error should probably be changed so that shorthand functions match the behavior of normal/arrow functions though.

**Comment by GlassBricks** [comment]
I did some investigating:

The context of functions are determined in `function-context.ts` by the _declaration_ of their inferred types, not by the types itself (I think this is so that @noSelf annotations can be looked up). This is currently not done with MethodDeclarations. I tried making it work with `MethodDeclaration`s, however, checker.getTypeAtNode() -> getAllCallSignatures -> getDeclaration() returns the `MethodDeclaration` itself, not the declaration of the type it was inferred from.
Reading the Type instead of the declaration would work if there is an explicit `this` parameter, but wouldn't work for @noSelf annotations.
How the type is inferred is buried deep inside typescript internal code...

Some things I might try to workaround this:
 - Trick typescript into treating method declarations as function properties instead
 - Get the type of the object literal instead, and index by the property name to get the method type
